### ISSUE TITLE
Conversion from bool to float sometimes produces tensor with values 255.0 for True instead of 1.0

### ISSUE THREAD

**Issue by matwilso** [pytorch/pytorch#54789]
## 🐛 Bug

When I have two seemingly identical bool tensors with values: `tensor([True, True, True, True])`, one of them produces a float tensor with values: `tensor([1.0, 1.0, 1.0, 1.0])`. The other one produces a float tensor with values: `tensor([255., 255., 255., 255.])`.

## To Reproduce

```python
import numpy as np
import torch
# these are the same data and they look the same when you load them, but they behave subtly differently
poisoned = np.frombuffer(b'\xff\xff\xff\xff', dtype=np.bool) # hex 255
clean = np.frombuffer(b'\x01\x01\x01\x01', dtype=np.bool)

torch_poisoned = torch.from_numpy(poisoned)
torch_clean = torch.from_numpy(clean)


print('poisoned == clean -->', (poisoned == clean).all())  # >>> True
print('torch_poisoned == torch_clean --> ', (torch_poisoned == torch_clean).all())  # >>> tensor(False)

print('torch_poisoned.dtype == torch_clean.dtype --> ', (torch_poisoned.dtype == torch_clean.dtype))  # >>> True
print('torch_poisoned.dtype, torch_clean.dtype --> ', torch_poisoned.dtype, torch_clean.dtype)  # >>> torch.bool, torch.bool

print('torch_clean: ',torch_clean)  # >>> tensor([True, True, True, True])
print('torch_clean.float(): ', torch_clean.float())  # >>> tensor([1., 1., 1., 1.])

print('torch_poisoned: ', torch_poisoned)  # >>> tensor([True, True, True, True])
print('torch_poisoned.float(): ', torch_poisoned.float())  # >>> tensor([255., 255., 255., 255.])
```

## Expected behavior

Any thing that torch calls a True bool should be converted to a float value of 1.0, not whatever bytes happened to be stored inside the bool.

This could just be an abuse of bool, but I came across this bug when using converting a boolean image using PIL. PyTorch should probably prevent users from being able to shoot their foot like this.

## Environment

Collecting environment information...
PyTorch version: 1.7.1+cu110
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.15.3

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.2.142
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 455.45.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.7.1+cu110
[pip3] torchaudio==0.7.2
[pip3] torchgeometry==0.1.2
[pip3] torchvision==0.8.2+cu110
[conda] Could not collect

## Additional context

cc @ezyang @gchanan @zou3519 @mruberry @rgommers @heitorschueroff

**Comment by mruberry** [comment]
NumPy will handle this case as expected:

```
poisoned.astype(np.float)
: array([1., 1., 1., 1.])
```

I think we would accept a PR fixing this as long as there wasn't a performance penalty.

**Comment by relh** [comment]
this also happens if you call `.int()`

**Comment by relh** [comment]
Also, a related issue:

```
(Pdb) torch_poisoned.to(0).max()
tensor(False, device='cuda:0')
(Pdb) torch_clean.to(0).max()
tensor(True, device='cuda:0')
```

When not on a GPU:
```
(Pdb) torch_poisoned.max()
tensor(True)
(Pdb) torch_clean.max()
tensor(True)
```

**Comment by relh** [comment]
This also impacts doing arithmetic with these poisoned tensors. 

These poisoned tensors exist in a fairly common use case. If you load PNG binary masks into boolean tensors and move them to GPUs they will act like the poisoned tensors above.

This happens even if you call `np.array` rather than `np.asarray` on the PIL Image

**Comment by jeromerony** [comment]
I can confirm that this also happens on 1.8.2 LTS.

It also happens when converting a PIL binary image to tensor.
```python
import numpy as np
import torch
from PIL import Image, ImageDraw

mask = Image.new(mode='1', size=(4, 4), color=False)
draw = ImageDraw.Draw(mask)
draw.polygon([(1, 1), (2, 1), (2, 2), (1, 2)], fill=True)

np_mask = np.array(mask)
print(np_mask)
tensor_mask = torch.from_numpy(np_mask)
print(tensor_mask) 
float_mask = tensor_mask.float()
print(float_mask)
```

**Comment by jeromerony** [comment]
It's also important to note that the behavior differs based on the device. On CPU, `True` values  will be converted to 255 while on a CUDA device, `True` will be converted to -1.
```python
import numpy as np
import torch

device = torch.device('cuda')
a = np.frombuffer(b'\xff\x00', dtype=bool)
print(a)                # >>> [ True False]
t = torch.from_numpy(a).to(device)
print(t)                # >>> tensor([ True, False], device='cuda:0')
print(t.float())        # >>> tensor([-1.,  0.], device='cuda:0')
```

**Comment by rfeinman** [comment]
I'm having this same issue, and I get different results on Linux vs. MacOS conda builds (both v1.10.1). It would be great to sort this out to foster reproducibility across different environments. It's especially important when using binary PIL images (`mode='1'`); in these cases `np.array(im)` yields a bool array with "True" bytes of value \xff).

Code snippet:
```python
a = np.frombuffer(b'\xff\x00', dtype=np.bool_)
t = torch.tensor(a, dtype=torch.float)
print(t)
```

Linux output:
```
tensor([255.,   0.])
```

MacOS output:
```
tensor([1., 0.])
```

**Comment by mruberry** [comment]
@rfeinman I don't immediately no, sorry. I had more context on this issue previously but I've forgotten most of it. My (very) vague recollection is that the C++ standard doesn't require the boolean `True` value be represented by 1 but any non-zero value, so different implementations may be doing different things while still being compliant. Still, even though the C++ standard may not guarantee this behavior it's clearly surprising to PyTorch users. Bumping priority based on user activity.

**Comment by ezyang** [comment]
Based on how many kernels are broken, it sounds like we are silently assuming that the values are only ever zero or one, but enforcement that this is the case is insufficient. But it's hard to see how we could enforce this, e.g., when receiving inputs from numpy, as numpy accepts "poisoned" boolean tensors and PyTorch by default just directly views into the Numpy memory. You'd have to do a full scan to check if it was bad or not and we don't want to do that. So it seems better for us to fix the kernels then.

**Comment by dtoniolo** [comment]
Found the same problem on Linux, but on macOS things work fine. Very annoying and possibly a deal breaker: finding all the possible places in which it may pop up in a codebase is a daunting, error prone task.

**Comment by philip-bl** [comment]
This issue is present on the current Google Colab version of pytorch. I stumbled upon it when I was converting binary PIL images to torch tensors. The version of pytorch is

```
Collecting environment information...
PyTorch version: 1.10.0+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0
Libc version: glibc-2.26

Python version: 3.7.13 (default, Mar 16 2022, 17:37:17)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.144+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.10.0+cu111
[pip3] torchaudio==0.10.0+cu111
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.11.0
[pip3] torchvision==0.11.1+cu111
[conda] Could not collect
```

**Comment by rems75** [comment]
This behavior is still happening in Pytorch 1.12.1. Is that expected? @peterbell10 @mruberry

**Comment by peterbell10** [comment]
The fix didn't land in time for the 1.12 release, but should be included in the upcoming 1.13 release.
